File: 1/elixir/santa_floor.exs
defmodule SantaFloor do
  @external_resource "../input.txt"

  def final_floor, do: hd floors

  def basement_reached do
    {_, index} = floors
    |> Enum.reverse
    |> Enum.with_index
    |> Enum.find(fn {floor, _index} -> floor == -1 end)
    index
  end

  defp input do
    {:ok, content} = File.read(@external_resource)
    content
  end

  defp floors do
    input
    |> String.codepoints
    |> Enum.reduce([0], fn(codepoint, [current | _tail] = floors) ->
      case codepoint do
        "(" -> [current + 1 | floors]
        ")" -> [current - 1 | floors]
        _   -> floors
      end
    end)
  end
end

IO.puts "Santa is going to floor #{SantaFloor.final_floor}"
IO.puts "Santa reaches the basement after #{SantaFloor.basement_reached} moves"

File: 1/ruby/santa_floor.rb

class SantaFloor
  UP = '('
  DOWN = ')'

  def initialize(input)
    @input = input
    validate_input
  end

  def final_floor
    floors.last
  end

  def basement_reached
    floors.index(-1)
  end

  private
  attr_reader :input

  def floors
    input.chars.reduce([0]) do |floors, direction|
      move(floors, direction)
    end
  end

  def move(floors, direction)
    case direction
    when UP
      floors << floors.last + 1
    when DOWN
      floors << floors.last - 1
    end
  end

  def validate_input
    unless input.delete('()').empty?
      fail 'Bad input. All characters must be ( or ).'
    end
  end
end

input = File.read('../input.txt').chomp
santa_floor = SantaFloor.new(input)
puts "Santa goes to floor #{santa_floor.final_floor}"
puts "Santa reaches the basement after #{santa_floor.basement_reached} moves"
